test/specs/wrapper/setProps.spec.js

```diff
@@ -187,6 +187,81 @@ describeWithShallowAndMount('setProps', mountingMethod => {
       await wrapper.setProps({ prop1 })
       expect(wrapper.vm.prop2).to.equal(prop1)
     })
+
+    it('invokes watchers with immediate set to "true"', async () => {
+      const callback = sinon.spy()
+      const TestComponent = {
+        template: '<div />',
+        props: ['propA'],
+        mounted() {
+          this.$watch(
+            'propA',
+            function() {
+              callback()
+            },
+            { immediate: true }
+          )
+        }
+      }
+      const wrapper = mountingMethod(TestComponent, {
+        propsData: { propA: 'none' }
+      })
+
+      expect(callback.calledOnce)
+      callback.resetHistory()
+
+      await wrapper.setProps({ propA: 'value' })
+      expect(wrapper.props().propA).to.equal('value')
+      expect(callback.calledOnce)
+    })
+
+    it('invokes watchers with immediate set to "true" with deep objects', async () => {
```

Comment:
also added a deep equality test here to test the assign by reference in the code to make sure we don't get into some type of continual recursive loop. This looks to work as intended